beacon-chain/cache/attestation.go

```diff
@@ -0,0 +1,262 @@
+	if local != nil {
+		if covered, err := att.GetAggregationBits().Contains(local.GetAggregationBits()); err != nil {
+			return err
+		} else if covered {
+			group.local = nil
+		}
+	}
+
+	attsToKeep := make([]ethpb.Att, 0, len(group.external))
+	for _, a := range group.external {
+		if covered, err := att.GetAggregationBits().Contains(a.GetAggregationBits()); err != nil {
+			return err
+		} else if !covered {
+			attsToKeep = append(attsToKeep, a)
+		}
+	}
+	group.external = attsToKeep
+
+	if group.local == nil && len(group.external) == 0 {
+		delete(c.attestations, id)
+	}
+
+	return nil
+}
+
+func (c *AttestationCache) PruneBefore(slot primitives.Slot) uint64 {
+	c.Lock()
+	defer c.Unlock()
+
+	var pruneCount uint64
+	for id, group := range c.attestations {
+		if group.slot < slot {
+			if group.local != nil {
+				pruneCount++
+			}
+			pruneCount += uint64(len(group.external))
+			delete(c.attestations, id)
+		}
+	}
+	return pruneCount
+}
+
+func (c *AttestationCache) AggregateIsRedundant(att ethpb.Att) (bool, error) {
+	if att.IsNil() {
+		return true, nil
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return true, errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.attestations[id]
+	if group == nil {
+		return false, nil
+	}
+
+	for _, a := range group.external {
+		if redundant, err := a.GetAggregationBits().Contains(att.GetAggregationBits()); err != nil {
+			return true, err
+		} else if redundant {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// SaveForkchoiceAttestation saves a forkchoice attestation.
+func (c *AttestationCache) SaveForkchoiceAttestation(att ethpb.Att) error {
+	return c.forkchoiceAtts.SaveForkchoiceAttestation(att)
+}
+
+// SaveForkchoiceAttestations saves forkchoice attestations.
+func (c *AttestationCache) SaveForkchoiceAttestations(att []ethpb.Att) error {
+	return c.forkchoiceAtts.SaveForkchoiceAttestations(att)
+}
+
+// ForkchoiceAttestations returns all forkchoice attestations.
+func (c *AttestationCache) ForkchoiceAttestations() []ethpb.Att {
+	return c.forkchoiceAtts.ForkchoiceAttestations()
+}
+
+// DeleteForkchoiceAttestation deletes a forkchoice attestation.
+func (c *AttestationCache) DeleteForkchoiceAttestation(att ethpb.Att) error {
+	return c.forkchoiceAtts.DeleteForkchoiceAttestation(att)
+}
+
+func GetBySlotAndCommitteeIndex[T ethpb.Att](c *AttestationCache, slot primitives.Slot, committeeIndex primitives.CommitteeIndex) []T {
+	c.RLock()
+	defer c.RUnlock()
+
+	var result []T
+
+	for _, group := range c.attestations {
+		local, ok := group.local.(T)
+		if ok {
+			if local.GetData().Slot == slot && local.CommitteeBitsVal().BitAt(uint64(committeeIndex)) {
+				result = append(result, local)
+				for _, a := range group.external {
+					a, ok := a.(T)
+					if ok {
+						result = append(result, a)
+					}
+				}
+			}
+		} else if len(group.external) > 0 {
+			a, ok := group.external[0].(T)
```

Comment:
We can safely compare the first attestation because all atts in a group must have the same slot and committee index, since they are under the same key